### PR TITLE
fix: widget CSP issues

### PR DESCRIPTION
**Pull request by aarongranick-okta**
## Description:

Is compliant with strict CSP policies. CSP policy is enforced in playground and in karma tests. Karma tests also enforce that all templates are precompiled.

Will resolve https://github.com/okta/okta-signin-widget/issues/408

Includes alpha version of Courage built from this branch: https://github.com/okta/okta-ui/pull/2127

The files being added under `packages/@okta/babel-plugin-handlebars-inline-precompile/` are copied from okta-ui/courage by the `courage-for-signin-widget` build script

## PR Checklist

- [x] Have you verified the basic functionality for this change?

- [x] Added unit tests?

- [ ] Added e2e tests

- [x] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

### Reviewers:
@haishengwu-okta 
@rajaguruduraisamy-okta 
@jmelberg-okta 

### Issue:

- [OKTA-286452](https://oktainc.atlassian.net/browse/OKTA-286452)

### CODE REVIEW

File: packages/@okta/courage-for-signin-widget/scripts/build.sh [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash -x
+
+# Copy handlebars precompile babel plugin
+mkdir -p ../babel-plugin-handlebars-inline-precompile
+cp -rf ./node_modules/@okta/babel-plugin-handlebars-inline-precompile/* ../babel-plugin-handlebars-inline-precompile/
```
Comment (haishengwu-okta) [inline packages/@okta/courage-for-signin-widget/scripts/build.sh:5]: does the webpack copy plugin support copy dir?

File: src/EnrollCallAndSmsController.js [hunk 2]
```diff
@@ -24,11 +27,13 @@ define([
 
   var _ = Okta._;
   var { Keys } = Okta.internal.util;
-  var warningTemplate = '<div class="okta-form-infobox-warning infobox infobox-warning login-timeout-warning">\
+  var warningTemplate = Okta.View.extend({
+    template: hbs('<div class="okta-form-infobox-warning infobox infobox-warning login-timeout-warning">\
```
Comment (haishengwu-okta) [inline src/EnrollCallAndSmsController.js:31]: nit:
shall we follow to avoid any DOM changes
```
View.extend({
  className: 'okta-form-infobox-warning....',
  template: hbs('<span...></span><p>..</p>')
})

File: src/util/Bundles.js [hunk 3]
```diff
@@ -141,10 +138,7 @@ define([
     // zh-CN -> zh_CN
     languageCode = language.replace('-', '_');
 
-    path = assets.rewrite(bundlePathTpl({
-      bundle: bundle,
-      languageCode: languageCode
-    }));
+    path = assets.rewrite(`/labels/json/${bundle}_${languageCode}.json`);
```
Comment (haishengwu-okta) [inline src/util/Bundles.js:141]: shall it run through `encodeURI` ?

File: src/v2/view-builder/internals/BaseView.js [hunk 4]
```diff
@@ -14,9 +15,12 @@ export default View.extend({
 
   className: 'siw-main-view',
 
-  template: '<div class="siw-main-header"></div>' +
-      '<div class="siw-main-body"></div>' +
-      '<div class="siw-main-footer"></div>',
+  template: hbs`
+    <div class="siw-main-header"></div>
+    <div class="siw-callout"></div>
```
Comment (haishengwu-okta) [inline src/v2/view-builder/internals/BaseView.js:20]: why add `siw-callout` ?

File: src/v2/view-builder/views/DeviceChallengePollView.js [hunk 5]
```diff
@@ -143,9 +146,13 @@ const Body = BaseForm.extend(Object.assign(
 
     doCustomURI () {
       this.ulDom && this.ulDom.remove();
-      this.ulDom = this.add(`
-        <iframe src="${this.customURI}" id="custom-uri-container" style="display:none;"></iframe>
-      `).last();
+      this.ulDom = this.add(View.extend({
+        template: hbs`<iframe src="{{{customURI}}}" id="custom-uri-container" style="display:none;"></iframe>`
```
Comment (haishengwu-okta) [inline src/v2/view-builder/views/DeviceChallengePollView.js:150]: just use the pure string as before? or does the `template` have to be handlerbar template now?

File: test/testcafe/spec/EnrollFactorPassword_spec.js [hunk 6]
```diff
@@ -31,16 +31,16 @@ test(`should have both password and confirmPassword fields and both are required
   // fields are required
   await enrollPasswordPage.clickNextButton();
   await enrollPasswordPage.waitForErrorBox();
-  await t.expect(enrollPasswordPage.getPasswordError()).eql('This field cannot be left blank');
-  await t.expect(enrollPasswordPage.getConfirmPasswordError()).eql('This field cannot be left blank');
+  await t.expect(enrollPasswordPage.getPasswordError()).contains('This field cannot be left blank');
```
Comment (haishengwu-okta) [inline test/testcafe/spec/EnrollFactorPassword_spec.js:34]: hmm, why cannot use `eql`?

File: test/unit/main.js [hunk 7]
```diff
@@ -1,3 +1,5 @@
+import './karma/karma-enforce-precompile';
```
Comment (haishengwu-okta) [inline test/unit/main.js:1]: you have added `customHeaders` in karma.conf. why still need this?

Comment (aarongranick-okta) [reply test/unit/main.js:1]: The custom headers will detect things like `eval` which we will not have as long as we are using `handlebars/runtime`. If we accidentally switch to full handlebars, the CSP will start failing again.

This check in karma detects a template that has not been precompiled. This would not throw a CSP error, instead the template would not be properly rendered. Depending on how the test is written we may not detect that or it may be confusing to the developer as to why the test is failing.

File: test/testcafe/spec/EnrollFactorPassword_spec.js (hunk 6 above)
Comment (aarongranick-okta) [reply test/testcafe/spec/EnrollFactorPassword_spec.js:34]: The precompiled templates sometimes have an extra whitespace in them (if the template in the sourcefile has line breaks and empty spaces). Most of these are stripped, but there are some cases where an extra space will stay in. (Input label is one of these cases)

File: src/v2/view-builder/views/DeviceChallengePollView.js (hunk 5 above)
Comment (aarongranick-okta) [reply src/v2/view-builder/views/DeviceChallengePollView.js:150]: At this point in time, we are allowing `add()` with a string. As long as it doesn't have handlebars in it, there is technically no problem with that. Before I added the karma enforce precompile check, I was converting some of these cases to be Views with templates because I had a warning for using `add()` with a bare string. In this case we can probably go back to pure string.

File: src/v2/view-builder/internals/BaseView.js (hunk 4 above)
Comment (aarongranick-okta) [reply src/v2/view-builder/internals/BaseView.js:20]: Good catch. I think this is probably a mistake from doing rebases/merges over the last couple months. I'll fix it.

File: test/testcafe/spec/EnrollFactorPassword_spec.js (hunk 6 above)
Comment (haishengwu-okta) [reply test/testcafe/spec/EnrollFactorPassword_spec.js:34]: `getPasswordError` tries to get inner text of following `p` element.
Wouldn't expect extra line break ? 
so maybe `trim` whilespace so that we can still use `eql` ?

```
<p id="input-container-error6" class="okta-form-input-error o-form-input-error o-form-explain" role="alert"><span class="icon icon-16 error-16-small" role="img" aria-label="Error"></span>This field cannot be left blank</p>
```

File: packages/@okta/courage-for-signin-widget/scripts/build.sh (hunk 1 above)
Comment (aarongranick-okta) [reply packages/@okta/courage-for-signin-widget/scripts/build.sh:5]: I was running into an issue using webpack to do the copy. This babel plugin is used BY the webpack at a very early phase, so the plugin needs to be in place before we start compiling any source files. There may be a way to get it to work. If this is important, I can spend some more time on it.

File: test/unit/main.js (hunk 7 above)
Comment (haishengwu-okta) [reply test/unit/main.js:1]: Then shall the changes made to framework itself (`courage-for-signin-widget/src/`)?

File: src/EnrollCallAndSmsController.js (hunk 2 above)
Comment (aarongranick-okta) [reply src/EnrollCallAndSmsController.js:31]: I'm not exactly sure how that works, this is related to how courage is implemented?

File: src/EnrollYubikeyController.js [hunk 8]
```diff
@@ -50,7 +51,9 @@ function (Okta, FormType, FormController, Footer, TextBox) {
       className: 'enroll-yubikey',
       formChildren: [
         FormType.View({
-          View: '<div class="yubikey-demo" data-type="yubikey-example"></div>'
+          View: Okta.View.extend({
+            template: hbs`<div class="yubikey-demo" data-type="yubikey-example"></div>`
```
Comment (aarongranick-okta) [inline src/EnrollYubikeyController.js:55]: note: may not need to be a handlebars template. I will check this.

File: src/util/Bundles.js (hunk 3 above)
Comment (aarongranick-okta) [reply src/util/Bundles.js:141]: It probably won't hurt. I'll give it a try.

File: test/testcafe/spec/EnrollFactorPassword_spec.js (hunk 6 above)
Comment (aarongranick-okta) [reply test/testcafe/spec/EnrollFactorPassword_spec.js:34]: I'll take another look at this

File: packages/@okta/courage-for-signin-widget/scripts/build.sh (hunk 1 above)
Comment (haishengwu-okta) [reply packages/@okta/courage-for-signin-widget/scripts/build.sh:5]: The problem is using wrapper build script is it doesn't support `webpack` parameters (although there is probably a quick and may not matter much for this package)
Would be good to avoid if possible (separate PR is fine).

File: src/EnrollCallAndSmsController.js (hunk 2 above)
Comment (haishengwu-okta) [reply src/EnrollCallAndSmsController.js:31]: yeah.
`View` will add another parent (`div` by default) around the `template`

File: test/testcafe/spec/EnrollFactorPassword_spec.js (hunk 6 above)
Comment (haishengwu-okta) [reply test/testcafe/spec/EnrollFactorPassword_spec.js:34]: can be separate PR if it's not quick fix.

File: webpack.test.config.js [hunk 9]
```diff
@@ -17,10 +17,10 @@ plugins.push(new webpack.DefinePlugin({
   SDK_VERSION: JSON.stringify(SDK_VERSION)
 }));
 
-testConfig.module.rules.push({
-  test: /\.js$/,
-  use: ['source-map-loader'],
-  enforce: 'pre'
+testConfig.module.rules.push({
```
Comment (haishengwu-okta) [inline webpack.test.config.js:20]: remove extra space

File: packages/@okta/courage-for-signin-widget/scripts/build.sh (hunk 1 above)
Comment (aarongranick-okta) [reply packages/@okta/courage-for-signin-widget/scripts/build.sh:5]: I see, you'd like to do something like `yarn build --watch` (instead of `npx webpack --watch`)?

File: test/unit/main.js (hunk 7 above)
Comment (aarongranick-okta) [reply test/unit/main.js:1]: I did originally have this check in CourageForSigninWidget.js, but I decided to move it to karma because I didn't want a runtime penalty of looking for `{{` in every template string, and I also want to be careful about causing a runtime error visible to users. Seeing a UI with `{{` in it is a lower priority bug than having an exception which breaks the whole login flow. Of course the karma check only works if templates are rendered within unit tests.  Another option would be an eslint  plugin to detect strings with `{{` that are not wrapped in `hbs`.

File: packages/@okta/courage-for-signin-widget/scripts/build.sh (hunk 1 above)
Comment (haishengwu-okta) [reply packages/@okta/courage-for-signin-widget/scripts/build.sh:5]: yeah and other parameters for `webpack` command.
it might not be big deal for this module but always trying to avoid build wrapper if possible.

File: test/unit/main.js (hunk 7 above)
Comment (haishengwu-okta) [reply test/unit/main.js:1]: `eslint plugin` sounds better!

File: test/testcafe/spec/EnrollFactorPassword_spec.js (hunk 6 above)
Comment (aarongranick-okta) [reply test/testcafe/spec/EnrollFactorPassword_spec.js:34]: I investigated, this is already fixed/tightened up on the courage side, so this must have been from an earlier iteration. I'll put it back to `eql`

File: eslint-local-rules.js [hunk 10]
```diff
@@ -0,0 +1,5 @@
+var noBareTemplates = require('./packages/@okta/eslint-plugin-okta-ui/lib/rules/no-bare-templates');
+
+module.exports = {
+  'no-bare-templates': noBareTemplates
```
Comment (haishengwu-okta) [inline eslint-local-rules.js:4]: 👍

File: packages/@okta/courage-for-signin-widget/scripts/build.sh (hunk 1 above)
Comment (aarongranick-okta) [reply packages/@okta/courage-for-signin-widget/scripts/build.sh]: I moved this to webpack copy and removed the build script

File: test/unit/main.js (hunk 7 above)
Comment (aarongranick-okta) [reply test/unit/main.js:1]: I updated the eslint plugin on the courage side and am copying it over with the build from courage-for-signin-widget.

File: packages/@okta/courage-for-signin-widget/package.json [hunk 11]
```diff
@@ -4,11 +4,12 @@
...
   "main": "src/CourageForSigninWidget.js",
   "scripts": {
-    "build": "scripts/build.sh"
+    "build": "webpack"
   },
   "devDependencies": {
-    "@okta/babel-plugin-handlebars-inline-precompile": "3.1.0-alpha.3586.gd86894f",
-    "@okta/courage": "4.6.0-alpha.3586.gd86894f",
+    "@okta/babel-plugin-handlebars-inline-precompile": "3.1.0-alpha.3660.g8faa6ac",
+    "@okta/eslint-plugin-okta-ui": "0.1.0-alpha.3660.g8faa6ac",
```
Comment (haishengwu-okta) [inline packages/@okta/courage-for-signin-widget/package.json:11]: what's `@okta/eslint-plugin-okta-ui` for?

Comment (aarongranick-okta) [reply packages/@okta/courage-for-signin-widget/package.json:11]: Thats where the `no-bare-templates` eslint rule is coming from. The same rule is run on courage

File: packages/@okta/courage-for-signin-widget/package.json [hunk 12]
```diff
@@ -4,11 +4,12 @@
   "description": "generates a Courage version for Sign-In Widget",
   "main": "src/CourageForSigninWidget.js",
   "scripts": {
-    "build": "scripts/build.sh"
+    "build": "webpack"
   },
   "devDependencies": {
-    "@okta/babel-plugin-handlebars-inline-precompile": "3.1.0-alpha.3586.gd86894f",
-    "@okta/courage": "4.6.0-alpha.3586.gd86894f",
+    "@okta/babel-plugin-handlebars-inline-precompile": "3.1.0-alpha.3660.g8faa6ac",
```
Comment (haishengwu-okta) [inline packages/@okta/courage-for-signin-widget/package.json:10]: This module `packages/@okta/courage-for-signin-widget/` getting more responsibility rather than just pulling Courage.
Some further refactoring seems necessary. (addressed separately) 
One idea:
- rename to more general module name like `okta-internal-modules`
- add script per module like
    - `build-courage`
    - `build-handlebar-inline-precompile`
    - `build-eslint-rule`
    - `build` compose all above (and apparently can be done in parallel)

**Comment by aarongranick-okta** [comment]
@haishengwu-okta Created new JIRA OKTA-311926 for `courage-for-signin-widget` refactor

File: packages/@okta/babel-plugin-handlebars-inline-precompile/handlebars/patch-precompile.js [hunk 13]
```diff
@@ -0,0 +1,17 @@
+// Monkey patch handlebars to remove extra whitespace from templates before they are precompiled
+module.exports = function (Handlebars) {
+  const precompile = Handlebars.precompile;
+  Handlebars.precompile = function (template) {
+    // Trim extra whitespace from template
+    template = template
+      .replace(/\s+</g, '<') // all spaces before opening tag
+      .replace(/>\s+/g, '>') // all spaces after closing tag
+      .replace(/}}\s+{{/g, '}}{{') // all spaces betwen mustaches
```
Comment (jmelberg-okta) [inline packages/@okta/babel-plugin-handlebars-inline-precompile/handlebars/patch-precompile.js:9]: Is this a safe operation? In admin we have many templates that rely on chaining one value to another with a space in between. Looks like the closest occurrence here (uses a slightly different format) is in [FactorsDropDownOptions.js](https://github.com/okta/okta-signin-widget/blob/d294c668f1a2fffb5078bd2f9ef72f39cb400d2b/src/views/mfa-verify/dropdown/FactorsDropDownOptions.js#L21).

If we omit this substitution, does anything break?

File: packages/@okta/courage-for-signin-widget/webpack.config.js [hunk 14]
```diff
@@ -6,10 +6,12 @@ const CopyWebpackPlugin = require('copy-webpack-plugin');
 const PACKAGE_JSON = require('./package.json');
 
 const EMPTY = resolve(__dirname, 'src/empty');
-const SHARED_JS = resolve(__dirname, 'node_modules/@okta/courage/src');
-const COURAGE_DIST = resolve(__dirname, 'node_modules/@okta/courage/dist');
-const PUBLISH_DIR = resolve(__dirname, '../courage-dist');
-const I18N_DIR = resolve(__dirname, '../i18n');
+const NODE_MODULES_SRC = resolve(__dirname, 'node_modules/@okta');
+const NODE_MODULES_DEST = resolve(__dirname, '..');
```
Comment (jmelberg-okta) [inline packages/@okta/courage-for-signin-widget/webpack.config.js:10]: super-nit: Switch the order of these vars to gradually increase:

```suggestion
const NODE_MODULES = resolve(__dirname, '..');
const OKTA_NODE_MODULES = resolve(NODE_MODULES_DEST, '@okta');
```

File: packages/@okta/eslint-plugin-okta-ui/lib/rules/no-bare-templates.js [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+  }
+  return false;
+}
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Detect use of bare (unwrapped) handlebars template strings',
+      category: 'CSP Issues',
+      recommended: true,
```
Comment (jmelberg-okta) [inline packages/@okta/eslint-plugin-okta-ui/lib/rules/no-bare-templates.js:33]: Would also be nice to provide logic for `--fix` since the change is minimal.

File: packages/@okta/babel-plugin-handlebars-inline-precompile/handlebars/patch-precompile.js (hunk 13 above)
Comment (aarongranick-okta) [reply packages/@okta/babel-plugin-handlebars-inline-precompile/handlebars/patch-precompile.js:9]: This file is copied from `okta-ui` by the `courage-for-signin-widget` build. There is a separate PR for these changes over here: https://github.com/okta/okta-ui/pull/2127

This prevents extra (unintentional) whitespace in the precompiled templates. Without this, the bundle size is larger and also many tests would need to be changed to account for extra whitespace.

File: packages/@okta/eslint-plugin-okta-ui/lib/rules/no-bare-templates.js (hunk 15 above)
Comment (aarongranick-okta) [reply packages/@okta/eslint-plugin-okta-ui/lib/rules/no-bare-templates.js:33]: Good suggestion. This file is also copied from `okta-ui`, the original source is in a separate PR: https://github.com/okta/okta-ui/pull/2127

**Comment by aarongranick-okta** [comment]
Changes can be tested in monolith using SIW version `4.2.0-alpha.gb8d39a3` https://bacon-go.aue1e.saasure.net/#!/commits/okta-signin-widget/branch/bacon-topic-ag-siw-csp

File: packages/@okta/babel-plugin-handlebars-inline-precompile/handlebars/patch-precompile.js (hunk 13 above)
Comment (jmelberg-okta) [reply packages/@okta/babel-plugin-handlebars-inline-precompile/handlebars/patch-precompile.js:9]: Both the changes proposed in [okta/okta-ui#2127](https://github.com/okta/okta-ui/pull/2127) and here have this potential side-effect.

In `okta-ui`, I'd assume we'd need to remove this line to make sure chained template values that expect spaces are met.

Examples:
- [PendingRequestsView.js](https://github.com/okta/okta-ui/blob/master/packages/admin/appinstance/src/assignments/views/PendingRequestsView.js#L72)
- [DevicesList.js](https://github.com/okta/okta-ui/blob/master/packages/admin/devices/src/views/DevicesList.js#L53)
- [UserPicker.js](https://github.com/okta/okta-ui/blob/master/packages/admin/shared/src/views/ud/mappings/components/UserPicker.js#L30)
- [SearchUtil](https://github.com/okta/okta-ui/blob/master/packages/admin/syslog/src/util/SearchUtil.js#L95)

How much extra of a bundle size does the removal of this space save? Refactoring our template logic to include spaces at the `*.properties` or passed param level (i.e `{ space: '  '}`) doesn't seem like the best option.

cc: @rajaguruduraisamy-okta

Comment (patrickmcdougle-okta) [reply packages/@okta/babel-plugin-handlebars-inline-precompile/handlebars/patch-precompile.js:9]: 👋🏻  Hi everyone, coming to you from the future as someone who has been tasked with updating SearchUtil.js to use precompiled templates and I just want to say that @jmelberg-okta was absolutely right this change is problematic there and the only way to proceed is to replace handlebars with es6 template literals.
